### PR TITLE
Isolation contexts

### PR DESCRIPTION
**Pull request by robdockins**

### CODE REVIEW

File: what4/src/What4/Config.hs [hunk 1]
```diff
@@ -890,28 +907,46 @@ instance Show OptCreateFailure where
...
+--   of the given options are already present in the configuration,
+--   nothing is done for that option and it is silently skipped.
+tryExtendConfig :: [ConfigDesc] -> Config -> IO ()
+tryExtendConfig ts (Config cfg) =
+  RWV.modify_ cfg (\m -> foldM tryInsertOption m ts)
+
+-- | Create a new configuration object that shares the option
+--   settings currently in the given input config. However,
+--   an options added to either configuration using @extendConfig@
+--   will not be propigated to the other .
```
Comment (kquick) [inline what4/src/What4/Config.hs:943]: "propagated", and an extra space before the period.

File: what4/src/What4/Config.hs [hunk 2]
```diff
@@ -890,28 +907,46 @@ instance Show OptCreateFailure where
...
+--
+--   To restate, option settings that already exist in the
```
Comment (kquick) [inline what4/src/What4/Config.hs:945]: Maybe "To restate" -> "Conversely"

File: what4/src/What4/Expr/Allocator.hs [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+import           What4.Config as CFG
+import           What4.Expr.App
+import           What4.ProgramLoc
+import           What4.Utils.AbstractDomains
+
+------------------------------------------------------------------------
+-- Cache start size
+
+-- | Starting size for element cache when caching is enabled.
+--
```
Comment (kquick) [inline what4/src/What4/Expr/Allocator.hs:40]: Size in .. ?  bytes?  elements?

I realize these were just moved, but seems like a good opportunity for clarification.

File: what4/src/What4/Expr/Allocator.hs [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+--   This option is named \"backend.cache_start_size\"
+cacheStartSizeOption :: ConfigOption BaseIntegerType
+cacheStartSizeOption = configOption BaseIntegerRepr "backend.cache_start_size"
+
+-- | The configuration option for setting the size of the initial hash set
+-- used by simple builder
```
Comment (kquick) [inline what4/src/What4/Expr/Allocator.hs:46]: Same issue: what units?

File: what4/src/What4/Expr/Builder.hs [hunk 5]
```diff
@@ -405,109 +386,69 @@ type instance SymExpr (ExprBuilder t st fs) = Expr t
...
+userState :: Lens' (ExprBuilder t st fs) (st t)
+userState = lens sbUserState (\sym st -> sym{ sbUserState = st })
+
+unaryThreshold :: Getter (ExprBuilder t st fs) (CFG.OptionSetting BaseIntegerType)
+unaryThreshold = to sbUnaryThreshold
+
+cacheStartSize :: Getter (ExprBuilder t st fs) (CFG.OptionSetting BaseIntegerType)
+cacheStartSize = to sbCacheStartSize
+
+-- | Return a new expr builder where the configuaration object has
```
Comment (kquick) [inline what4/src/What4/Expr/Builder.hs:401]: "configuration"

File: what4/CHANGES.md [hunk 6]
```diff
@@ -1,5 +1,9 @@
 # next (TBA)
 
+* Add operations for increased control over the scope of
+  configuration options, both in the `What4.Confg` and
```
Comment (kquick) [inline what4/CHANGES.md:4]: "`What4.Config`"

**Review by kquick — APPROVED** [review]
Looks good; just a few typos and clarifications in comments/doc.

File: what4/src/What4/Expr/Builder.hs [hunk 7]
```diff
@@ -686,21 +682,6 @@ unaryThresholdDesc = CFG.mkOpt unaryThresholdOption sty help (Just (ConcreteInte
...
         help = Just "Maximum number of values in unary bitvector encoding."
 
-------------------------------------------------------------------------
--- Configuration option for controlling how many disjoint ranges
--- should be allowed in bitvector domains.
-
--- | Maximum number of ranges in bitvector abstract domains.
---
---   This option is named \"backend.bvdomain_range_limit\"
-bvdomainRangeLimitOption :: CFG.ConfigOption BaseIntegerType
```
Comment (RyanGlScott) [inline what4/src/What4/Expr/Builder.hs:696]: Mention `bvdomainRangeLimitOption`'s removal in the changelog for good measure.

File: what4/src/What4/Expr/Allocator.hs [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    Just e -> return e
+    Nothing -> do
+      n <- freshNonce g
+      let e = NonceAppExpr $ NonceAppExprCtor { nonceExprId = n
+                                            , nonceExprLoc = pc
+                                            , nonceExprApp = p
+                                            , nonceExprAbsValue = v
+                                            }
+      seq e $ stToIO $ PH.insert h p e
+      return $! e
```
Comment (RyanGlScott) [inline what4/src/What4/Expr/Allocator.hs:166]: Is the use of `$!` here actually necessary? The line above it calls `seq e`, after all.

File: what4/src/What4/FloatMode.hs [hunk 9]
```diff
@@ -0,0 +1,61 @@
+-----------------------------------------------------------------------
+-- |
+-- Module           : What4.FloatMode
+-- Description      :
```
Comment (RyanGlScott) [inline what4/src/What4/FloatMode.hs:4]: A brief description here would be helpful.

File: what4/src/What4/Config.hs [hunk 10]
```diff
@@ -890,28 +907,46 @@ instance Show OptCreateFailure where
...
+-- | Extend an existing configuration with new options. If any
+--   of the given options are already present in the configuration,
+--   nothing is done for that option and it is silently skipped.
+tryExtendConfig :: [ConfigDesc] -> Config -> IO ()
+tryExtendConfig ts (Config cfg) =
+  RWV.modify_ cfg (\m -> foldM tryInsertOption m ts)
+
+-- | Create a new configuration object that shares the option
+--   settings currently in the given input config. However,
+--   an options added to either configuration using @extendConfig@
```
Comment (RyanGlScott) [inline what4/src/What4/Config.hs:942]:
```suggestion
--   any options added to either configuration using @extendConfig@
```

File: what4/src/What4/Expr/Builder.hs [hunk 11]
```diff
@@ -384,6 +386,69 @@ type instance SymExpr (ExprBuilder t st fs) = Expr t
 type instance BoundVar (ExprBuilder t st fs) = ExprBoundVar t
 type instance SymAnnotation (ExprBuilder t st fs) = Nonce t
 
+exprCounter :: Getter (ExprBuilder t st fs) (NonceGenerator IO t)
```
Comment (RyanGlScott) [inline what4/src/What4/Expr/Builder.hs:389]: The type of `exprCounter` has changed from an ordinary field selector to a `Getter`, which is worth mentioning in the changelog. Similarly for `cacheStartSize` and `userState`.

File: what4/src/What4/Expr/Allocator.hs (hunk 8 above)
Comment (robdockins) [reply what4/src/What4/Expr/Allocator.hs:166]: good question; I don't think so

File: what4/CHANGES.md [hunk 12]
```diff
@@ -1,5 +1,19 @@
 # next (TBA)
 
+* Add operations for increased control over the scope of
+  configuration options, both in the `What4.Config` and
+  `What4.Expr.Builder` modules.
+
+* Previously, the `exprCounter`, `sbUserState`, `sbUnaryThreshold`, and
+  `sbCacheStartSize` fields of `ExprBuilder` were directly exposed; in
+  principle this allows users to modify them, which was not intended
+  in some cases.  These have been uniformy renamed to remove the `sb`
```
Comment (RyanGlScott) [inline what4/CHANGES.md:10]:
```suggestion
  in some cases.  These have been uniformly renamed to remove the `sb`
```

**Review by RyanGlScott — APPROVED** [review]

**Pull request merged**